File: tools/export-to-parquet.py
#!/usr/bin/env python3
"""Export bench dataset from msgpack.gz → Parquet files.

Memory-optimized: uses msgpack streaming to extract each section independently,
never holding the full dataset in memory at once.

Usage:
    cd lib/shgat-tf
    python3 tools/export-to-parquet.py [--data-path <path>]
"""

import gc
import gzip
import json
import os
import struct
import sys
import time

import msgpack
import pyarrow as pa
import pyarrow.parquet as pq

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
GRU_DATA_DIR = os.path.normpath(os.path.join(SCRIPT_DIR, "../../gru/data"))


def get_arg(name, default):
    try:
        idx = sys.argv.index(f"--{name}")
        return sys.argv[idx + 1] if idx + 1 < len(sys.argv) else default
    except ValueError:
        return default


def emb_bytes(emb):
    return struct.pack(f"{len(emb)}f", *emb)


def sparse_bytes(sparse):
    if not sparse:
        return b"", b""
    return (
        struct.pack(f"{len(sparse)}i", *(int(s[0]) for s in sparse)),
        struct.pack(f"{len(sparse)}f", *(float(s[1]) for s in sparse)),
    )


def rss_mb():
    try:
        with open("/proc/self/status") as f:
            for line in f:
                if line.startswith("VmRSS:"):
                    return int(line.split()[1]) / 1024
    except Exception:
        pass
    return 0


def extract_section(raw_bytes, key):
    """Extract a single top-level key from msgpack bytes.

    The dataset is a map: {nodes: [...], leafIds: [...], ...}
    We decode the full map but immediately extract and return only the requested key,
    then delete the full map.
    """
    # For msgpack maps, we can't easily stream by key without custom parsing.
    # Instead, use unpackb with a hook that skips large arrays we don't need.
    # Actually, msgpack doesn't support random access, so we decode fully
    # but use raw=True for bytes efficiency and only convert what we need.
    pass


def main():
    data_path = get_arg("data-path",
                        os.path.join(GRU_DATA_DIR, "bench-dataset-export.msgpack.gz"))
    print("=== Export Parquet from msgpack.gz (Python streaming) ===\n")

    # --- Decompress (keep raw bytes, don't decode yet) ---
    print(f"[Data] Decompressing {data_path}...")
    t0 = time.time()
    with gzip.open(data_path, "rb") as f:
        raw = f.read()
    print(f"  {len(raw) / 1e6:.1f}MB decompressed in {time.time() - t0:.1f}s (RSS: {rss_mb():.0f}MB)")

    # --- Decode with list_hook to use tuples (less memory than lists) ---
    print("  Decoding msgpack (raw=True for memory efficiency)...")
    t1 = time.time()
    # raw=False needed for string keys, but we can use max_buffer_size
    unpacker = msgpack.Unpacker(max_buffer_size=2 * 1024 * 1024 * 1024)
    unpacker.feed(raw)
    ds = unpacker.unpack()
    del raw, unpacker
    gc.collect()
    print(f"  Decoded in {time.time() - t1:.1f}s (RSS: {rss_mb():.0f}MB)")

    # Quick stats
    nodes = ds[b"nodes"] if b"nodes" in ds else ds.get("nodes", [])
    leaf_ids = ds[b"leafIds"] if b"leafIds" in ds else ds.get("leafIds", [])
    emb_dim = ds[b"embeddingDim"] if b"embeddingDim" in ds else ds.get("embeddingDim", 0)
    prod_train = ds[b"prodTrain"] if b"prodTrain" in ds else ds.get("prodTrain", [])
    prod_test = ds[b"prodTest"] if b"prodTest" in ds else ds.get("prodTest", [])
    n8n_train = ds[b"n8nTrain"] if b"n8nTrain" in ds else ds.get("n8nTrain", [])
    n8n_eval = ds[b"n8nEval"] if b"n8nEval" in ds else ds.get("n8nEval", [])
    wf_tool_lists = ds[b"workflowToolLists"] if b"workflowToolLists" in ds else ds.get("workflowToolLists", [])

    def s(x):
        return x.decode() if isinstance(x, bytes) else x

    print(f"  Nodes: {len(nodes)} ({len(leaf_ids)} leaves), EmbDim: {emb_dim}")
    print(f"  Prod: {len(prod_train)} train / {len(prod_test)} test")
    print(f"  N8n: {len(n8n_train)} train / {len(n8n_eval)} eval")
    print(f"  RSS: {rss_mb():.0f}MB\n")

    counts = {
        "nodes": len(nodes),
        "prodTrain": len(prod_train),
        "prodTest": len(prod_test),
        "n8nTrain": len(n8n_train),
        "n8nEval": len(n8n_eval),
    }

    # ======== NODES ========
    t = time.time()
    print("[Parquet] Nodes...")
    k_id = b"id" if b"id" in (nodes[0] if nodes else {}) else "id"
    k_emb = b"embedding" if b"embedding" in (nodes[0] if nodes else {}) else "embedding"
    k_ch = b"children" if b"children" in (nodes[0] if nodes else {}) else "children"
    k_lv = b"level" if b"level" in (nodes[0] if nodes else {}) else "level"

    table = pa.table({
        "id": pa.array([s(n[k_id]) for n in nodes], pa.string()),
        "embedding": pa.array([emb_bytes(n[k_emb]) for n in nodes], pa.binary()),
        "children_json": pa.array([json.dumps([s(c) for c in n[k_ch]]) for n in nodes], pa.string()),
        "level": pa.array([n[k_lv] for n in nodes], pa.int32()),
    })
    out = os.path.join(GRU_DATA_DIR, "bench-nodes.parquet")
    pq.write_table(table, out, compression="snappy")
    print(f"  {table.num_rows} rows → {os.path.getsize(out)/1e6:.1f}MB ({time.time()-t:.1f}s, RSS: {rss_mb():.0f}MB)")
    del table
    # Free nodes from the dict
    if b"nodes" in ds: ds[b"nodes"] = None
    elif "nodes" in ds: ds["nodes"] = None
    del nodes; gc.collect()

    # ======== PROD TRAIN ========
    t = time.time()
    print("[Parquet] Prod train...")
    ex = prod_train
    k_ie = b"intentEmbedding" if ex and b"intentEmbedding" in ex[0] else "intentEmbedding"
    k_ct = b"contextToolIds" if ex and b"contextToolIds" in ex[0] else "contextToolIds"
    k_tt = b"targetToolId" if ex and b"targetToolId" in ex[0] else "targetToolId"
    k_it = b"isTerminal" if ex and b"isTerminal" in ex[0] else "isTerminal"
    k_tr = b"_traceId" if ex and b"_traceId" in ex[0] else "_traceId"

    table = pa.table({
        "intent_embedding": pa.array([emb_bytes(e[k_ie]) for e in ex], pa.binary()),
        "context_tool_ids_json": pa.array([json.dumps([s(c) for c in e[k_ct]]) for e in ex], pa.string()),
        "target_tool_id": pa.array([s(e[k_tt]) for e in ex], pa.string()),
        "is_terminal": pa.array([e[k_it] for e in ex], pa.int32()),
        "trace_id": pa.array([s(e[k_tr]) for e in ex], pa.string()),
    })
    out = os.path.join(GRU_DATA_DIR, "bench-prod-train.parquet")
    pq.write_table(table, out, compression="snappy")
    print(f"  {table.num_rows} rows → {os.path.getsize(out)/1e6:.1f}MB ({time.time()-t:.1f}s)")
    del table, ex
    if b"prodTrain" in ds: ds[b"prodTrain"] = None
    elif "prodTrain" in ds: ds["prodTrain"] = None
    del prod_train; gc.collect()

    # ======== PROD TEST ========
    t = time.time()
    print("[Parquet] Prod test...")
    ex = prod_test
    table = pa.table({
        "intent_embedding": pa.array([emb_bytes(e[k_ie]) for e in ex], pa.binary()),
        "context_tool_ids_json": pa.array([json.dumps([s(c) for c in e[k_ct]]) for e in ex], pa.string()),
        "target_tool_id": pa.array([s(e[k_tt]) for e in ex], pa.string()),
        "is_terminal": pa.array([e[k_it] for e in ex], pa.int32()),
        "trace_id": pa.array([s(e[k_tr]) for e in ex], pa.string()),
    })
    out = os.path.join(GRU_DATA_DIR, "bench-prod-test.parquet")
    pq.write_table(table, out, compression="snappy")
    print(f"  {table.num_rows} rows → {os.path.getsize(out)/1e6:.1f}MB ({time.time()-t:.1f}s)")
    del table, ex
    if b"prodTest" in ds: ds[b"prodTest"] = None
    elif "prodTest" in ds: ds["prodTest"] = None
    del prod_test; gc.collect()

    # ======== N8N TRAIN ========
    t = time.time()
    print("[Parquet] N8n train...")
    ex = n8n_train
    k_st = b"softTargetSparse" if ex and b"softTargetSparse" in ex[0] else "softTargetSparse"
    idx_b, prb_b = [], []
    for e in ex:
        ib, pb = sparse_bytes(e.get(k_st, []))
        idx_b.append(ib); prb_b.append(pb)
    table = pa.table({
        "intent_embedding": pa.array([emb_bytes(e[k_ie]) for e in ex], pa.binary()),
        "context_tool_ids_json": pa.array([json.dumps([s(c) for c in e[k_ct]]) for e in ex], pa.string()),
        "target_tool_id": pa.array([s(e[k_tt]) for e in ex], pa.string()),
        "is_terminal": pa.array([e[k_it] for e in ex], pa.int32()),
        "soft_target_indices": pa.array(idx_b, pa.binary()),
        "soft_target_probs": pa.array(prb_b, pa.binary()),
    })
    out = os.path.join(GRU_DATA_DIR, "bench-n8n-train.parquet")
    pq.write_table(table, out, compression="snappy")
    print(f"  {table.num_rows} rows → {os.path.getsize(out)/1e6:.1f}MB ({time.time()-t:.1f}s)")
    del table, ex, idx_b, prb_b
    if b"n8nTrain" in ds: ds[b"n8nTrain"] = None
    elif "n8nTrain" in ds: ds["n8nTrain"] = None
    del n8n_train; gc.collect()

    # ======== N8N EVAL ========
    t = time.time()
    print("[Parquet] N8n eval...")
    ex = n8n_eval
    idx_b, prb_b = [], []
    for e in ex:
        ib, pb = sparse_bytes(e.get(k_st, []))
        idx_b.append(ib); prb_b.append(pb)
    table = pa.table({
        "intent_embedding": pa.array([emb_bytes(e[k_ie]) for e in ex], pa.binary()),
        "context_tool_ids_json": pa.array([json.dumps([s(c) for c in e[k_ct]]) for e in ex], pa.string()),
        "target_tool_id": pa.array([s(e[k_tt]) for e in ex], pa.string()),
        "is_terminal": pa.array([e[k_it] for e in ex], pa.int32()),
        "soft_target_indices": pa.array(idx_b, pa.binary()),
        "soft_target_probs": pa.array(prb_b, pa.binary()),
    })
    out = os.path.join(GRU_DATA_DIR, "bench-n8n-eval.parquet")
    pq.write_table(table, out, compression="snappy")
    print(f"  {table.num_rows} rows → {os.path.getsize(out)/1e6:.1f}MB ({time.time()-t:.1f}s)")
    del table, ex, idx_b, prb_b
    if b"n8nEval" in ds: ds[b"n8nEval"] = None
    elif "n8nEval" in ds: ds["n8nEval"] = None
    del n8n_eval; gc.collect()

    # ======== METADATA ========
    meta = {
        "leafIds": [s(x) for x in leaf_ids],
        "embeddingDim": emb_dim,
        "workflowToolLists": [[s(t) for t in wf] for wf in wf_tool_lists],
        "exportedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "source": "export-to-parquet.py",
        "counts": counts,
    }
    meta_path = os.path.join(GRU_DATA_DIR, "bench-metadata.json")
    with open(meta_path, "w") as f:
        json.dump(meta, f)
    print(f"\n[Metadata] bench-metadata.json ({os.path.getsize(meta_path)/1024:.1f}KB)")

    print(f"\n=== Export complete (RSS: {rss_mb():.0f}MB) ===")
    print("Files:")
    for fn in ["bench-nodes.parquet", "bench-prod-train.parquet", "bench-prod-test.parquet",
               "bench-n8n-train.parquet", "bench-n8n-eval.parquet", "bench-metadata.json"]:
        fp = os.path.join(GRU_DATA_DIR, fn)
        if os.path.exists(fp):
            print(f"  {fn}: {os.path.getsize(fp)/1e6:.1f}MB")


if __name__ == "__main__":
    main()

File: tools/export-to-parquet.ts
#!/usr/bin/env -S deno run --allow-read --allow-write --allow-env --allow-net
/**
 * Export bench dataset from msgpack.gz → Parquet files.
 *
 * Reads the existing bench-dataset-export.msgpack.gz and writes:
 *   bench-nodes.parquet       — nodes (id, embedding bytes, children JSON, level)
 *   bench-prod-train.parquet  — prod training examples
 *   bench-prod-test.parquet   — prod test examples
 *   bench-n8n-train.parquet   — n8n training examples
 *   bench-n8n-eval.parquet    — n8n eval examples
 *   bench-metadata.json       — small JSON with leafIds, embeddingDim, workflowToolLists
 *
 * Embeddings are stored as raw Float32 bytes in a Binary column (1024 * 4 = 4096 bytes/row).
 * String arrays (children, contextToolIds) are stored as JSON-encoded strings.
 * softTargetSparse indices/probs are stored as separate Binary columns.
 *
 * Usage:
 *   cd lib/shgat-tf
 *   deno run --allow-read --allow-write --allow-env --allow-net \
 *     tools/export-to-parquet.ts [--data-path <path>]
 *
 * @module shgat-tf/tools/export-to-parquet
 */

import { dirname, resolve, fromFileUrl } from "https://deno.land/std@0.224.0/path/mod.ts";
import { decode as msgpackDecode } from "@msgpack/msgpack";
import pako from "pako";
import * as arrow from "apache-arrow";
// parquet-wasm: Node entry point auto-initializes WASM (no initWasm needed).
// ESM entry point requires calling the default export first.
// Deno npm: resolves to the Node entry point.
import parquetWasmModule from "parquet-wasm";
import {
  Compression,
  Table as WasmTable,
  writeParquet,
  WriterPropertiesBuilder,
} from "parquet-wasm";

// ==========================================================================
// Types (same as train-ob.ts)
// ==========================================================================

interface ExportedNode {
  id: string;
  embedding: number[];
  children: string[];
  level: number;
}

interface ProdExample {
  intentEmbedding: number[];
  contextToolIds: string[];
  targetToolId: string;
  isTerminal: number;
  _traceId: string;
}

interface N8nExample {
  intentEmbedding: number[];
  contextToolIds: string[];
  targetToolId: string;
  isTerminal: number;
  softTargetSparse: [number, number][];
}

interface ExportedDataset {
  nodes: ExportedNode[];
  leafIds: string[];
  embeddingDim: number;
  workflowToolLists: string[][];
  prodTrain: ProdExample[];
  prodTest: ProdExample[];
  n8nTrain: N8nExample[];
  n8nEval: N8nExample[];
}

// ==========================================================================
// CLI
// ==========================================================================

const cliArgs = Deno.args;

function getArg(name: string, def: string): string {
  const idx = cliArgs.indexOf(`--${name}`);
  if (idx === -1) return def;
  const next = cliArgs[idx + 1];
  return next && !next.startsWith("--") ? next : def;
}

const scriptDir = dirname(fromFileUrl(import.meta.url));
const GRU_DATA_DIR = resolve(scriptDir, "../../gru/data");
const dataPath = getArg("data-path", resolve(GRU_DATA_DIR, "bench-dataset-export.msgpack.gz"));

if (cliArgs.includes("--help")) {
  console.log(`
Export bench dataset from msgpack.gz to Parquet files.

Options:
  --data-path <path>   Path to msgpack.gz dataset (default: auto)
  --help               Show this help
`);
  Deno.exit(0);
}

// ==========================================================================
// Helpers
// ==========================================================================

/**
 * Convert a number[] embedding to raw Float32 bytes.
 * Returns a Uint8Array of length embDim * 4.
 */
function embeddingToBytes(emb: number[]): Uint8Array {
  const f32 = new Float32Array(emb);
  return new Uint8Array(f32.buffer);
}

/**
 * Convert a sparse target array [[idx, prob], ...] to two separate byte arrays.
 * Indices: Int32Array bytes. Probs: Float32Array bytes.
 */
function softTargetToBytes(sparse: [number, number][]): { indicesBytes: Uint8Array; probsBytes: Uint8Array } {
  if (!sparse || sparse.length === 0) {
    return { indicesBytes: new Uint8Array(0), probsBytes: new Uint8Array(0) };
  }
  const indices = new Int32Array(sparse.map(([idx]) => idx));
  const probs = new Float32Array(sparse.map(([, prob]) => prob));
  return {
    indicesBytes: new Uint8Array(indices.buffer),
    probsBytes: new Uint8Array(probs.buffer),
  };
}

/**
 * Write an Arrow Table to a Parquet file using parquet-wasm with Snappy compression.
 */
function writeParquetFile(table: arrow.Table, filePath: string): void {
  const ipcStream = arrow.tableToIPC(table, "stream");
  const wasmTable = WasmTable.fromIPCStream(ipcStream);
  const writerProps = new WriterPropertiesBuilder()
    .setCompression(Compression.SNAPPY)
    .build();
  const parquetBytes = writeParquet(wasmTable, writerProps);
  Deno.writeFileSync(filePath, parquetBytes);
}

// ==========================================================================
// Build Arrow tables
// ==========================================================================

/**
 * Build nodes table: id (Utf8), embedding (Binary), children (Utf8 JSON), level (Int32)
 */
function buildNodesTable(nodes: ExportedNode[]): arrow.Table {
  const n = nodes.length;
  const ids: string[] = new Array(n);
  const levels = new Int32Array(n);
  const childrenJson: string[] = new Array(n);
  // For embeddings, we'll build a Binary vector via makeBuilder
  const embBuilderType = new arrow.Binary();
  const embBuilder = arrow.makeBuilder({ type: embBuilderType, nullValues: [null] });

  for (let i = 0; i < n; i++) {
    ids[i] = nodes[i].id;
    levels[i] = nodes[i].level;
    childrenJson[i] = JSON.stringify(nodes[i].children);
    embBuilder.append(embeddingToBytes(nodes[i].embedding));
  }
  embBuilder.finish();
  const embVector = embBuilder.toVector();

  // Build table from columns
  const idVector = arrow.vectorFromArray(ids, new arrow.Utf8());
  const levelVector = arrow.makeVector(levels);
  const childrenVector = arrow.vectorFromArray(childrenJson, new arrow.Utf8());

  return new arrow.Table({
    id: idVector,
    embedding: embVector,
    children_json: childrenVector,
    level: levelVector,
  });
}

/**
 * Build prod examples table:
 *   intent_embedding (Binary), context_tool_ids (Utf8 JSON),
 *   target_tool_id (Utf8), is_terminal (Int32), trace_id (Utf8)
 */
function buildProdTable(examples: ProdExample[]): arrow.Table {
  const n = examples.length;
  const targetToolIds: string[] = new Array(n);
  const isTerminals = new Int32Array(n);
  const traceIds: string[] = new Array(n);
  const contextToolIdsJson: string[] = new Array(n);

  const embBuilderType = new arrow.Binary();
  const embBuilder = arrow.makeBuilder({ type: embBuilderType, nullValues: [null] });

  for (let i = 0; i < n; i++) {
    const ex = examples[i];
    embBuilder.append(embeddingToBytes(ex.intentEmbedding));
    contextToolIdsJson[i] = JSON.stringify(ex.contextToolIds);
    targetToolIds[i] = ex.targetToolId;
    isTerminals[i] = ex.isTerminal;
    traceIds[i] = ex._traceId;
  }
  embBuilder.finish();

  return new arrow.Table({
    intent_embedding: embBuilder.toVector(),
    context_tool_ids_json: arrow.vectorFromArray(contextToolIdsJson, new arrow.Utf8()),
    target_tool_id: arrow.vectorFromArray(targetToolIds, new arrow.Utf8()),
    is_terminal: arrow.makeVector(isTerminals),
    trace_id: arrow.vectorFromArray(traceIds, new arrow.Utf8()),
  });
}

/**
 * Build n8n examples table:
 *   intent_embedding (Binary), context_tool_ids (Utf8 JSON),
 *   target_tool_id (Utf8), is_terminal (Int32),
 *   soft_target_indices (Binary), soft_target_probs (Binary)
 */
function buildN8nTable(examples: N8nExample[]): arrow.Table {
  const n = examples.length;
  const targetToolIds: string[] = new Array(n);
  const isTerminals = new Int32Array(n);
  const contextToolIdsJson: string[] = new Array(n);

  const embType = new arrow.Binary();
  const embBuilder = arrow.makeBuilder({ type: embType, nullValues: [null] });
  const indicesBuilder = arrow.makeBuilder({ type: new arrow.Binary(), nullValues: [null] });
  const probsBuilder = arrow.makeBuilder({ type: new arrow.Binary(), nullValues: [null] });

  for (let i = 0; i < n; i++) {
    const ex = examples[i];
    embBuilder.append(embeddingToBytes(ex.intentEmbedding));
    contextToolIdsJson[i] = JSON.stringify(ex.contextToolIds);
    targetToolIds[i] = ex.targetToolId;
    isTerminals[i] = ex.isTerminal;

    const { indicesBytes, probsBytes } = softTargetToBytes(ex.softTargetSparse);
    indicesBuilder.append(indicesBytes);
    probsBuilder.append(probsBytes);
  }
  embBuilder.finish();
  indicesBuilder.finish();
  probsBuilder.finish();

  return new arrow.Table({
    intent_embedding: embBuilder.toVector(),
    context_tool_ids_json: arrow.vectorFromArray(contextToolIdsJson, new arrow.Utf8()),
    target_tool_id: arrow.vectorFromArray(targetToolIds, new arrow.Utf8()),
    is_terminal: arrow.makeVector(isTerminals),
    soft_target_indices: indicesBuilder.toVector(),
    soft_target_probs: probsBuilder.toVector(),
  });
}

// ==========================================================================
// MAIN
// ==========================================================================

console.log("=== Export Parquet from msgpack.gz ===\n");

// ---- Init parquet-wasm ----
// Node entry point: auto-initialized (parquetWasmModule is the module itself).
// ESM entry point: parquetWasmModule is the async init function.
// We call it if it's a function (ESM path), otherwise it's already initialized (Node path).
console.log("[WASM] Initializing parquet-wasm...");
if (typeof parquetWasmModule === "function") {
  await (parquetWasmModule as unknown as () => Promise<void>)();
}
console.log("[WASM] Ready.\n");

// ---- Load msgpack.gz dataset (staged to minimize peak memory) ----
console.log(`[Data] Loading ${dataPath}...`);
let ds: ExportedDataset;
{
  let compressed: Uint8Array | null = Deno.readFileSync(dataPath);
  console.log(`  Compressed: ${(compressed.byteLength / 1e6).toFixed(1)}MB`);
  let raw: Uint8Array | null = pako.ungzip(compressed);
  compressed = null; // free compressed buffer
  console.log(`  Decompressed: ${(raw.byteLength / 1e6).toFixed(1)}MB`);
  ds = msgpackDecode(raw) as ExportedDataset;
  raw = null; // free decompressed buffer
}

console.log(`  Nodes: ${ds.nodes.length} (${ds.leafIds.length} leaves), EmbDim: ${ds.embeddingDim}`);
console.log(`  Prod: ${ds.prodTrain.length} train / ${ds.prodTest.length} test`);
console.log(`  N8n: ${ds.n8nTrain.length} train / ${ds.n8nEval.length} eval\n`);

// ---- Write nodes ----
{
  const t0 = performance.now();
  console.log("[Parquet] Building nodes table...");
  const table = buildNodesTable(ds.nodes);
  console.log(`  Schema: ${table.schema.fields.map((f: { name: string; type: unknown }) => `${f.name}:${f.type}`).join(", ")}`);
  console.log(`  Rows: ${table.numRows}`);

  const outPath = resolve(GRU_DATA_DIR, "bench-nodes.parquet");
  writeParquetFile(table, outPath);
  const stat = Deno.statSync(outPath);
  console.log(`  Written: ${outPath} (${(stat.size / 1e6).toFixed(1)}MB, ${(performance.now() - t0).toFixed(0)}ms)\n`);
}

// ---- Write prod train ----
{
  const t0 = performance.now();
  console.log("[Parquet] Building prod-train table...");
  const table = buildProdTable(ds.prodTrain);
  console.log(`  Rows: ${table.numRows}`);

  const outPath = resolve(GRU_DATA_DIR, "bench-prod-train.parquet");
  writeParquetFile(table, outPath);
  const stat = Deno.statSync(outPath);
  console.log(`  Written: ${outPath} (${(stat.size / 1e6).toFixed(1)}MB, ${(performance.now() - t0).toFixed(0)}ms)\n`);
}

// ---- Write prod test ----
{
  const t0 = performance.now();
  console.log("[Parquet] Building prod-test table...");
  const table = buildProdTable(ds.prodTest);
  console.log(`  Rows: ${table.numRows}`);

  const outPath = resolve(GRU_DATA_DIR, "bench-prod-test.parquet");
  writeParquetFile(table, outPath);
  const stat = Deno.statSync(outPath);
  console.log(`  Written: ${outPath} (${(stat.size / 1e6).toFixed(1)}MB, ${(performance.now() - t0).toFixed(0)}ms)\n`);
}

// ---- Write n8n train ----
{
  const t0 = performance.now();
  console.log("[Parquet] Building n8n-train table...");
  const table = buildN8nTable(ds.n8nTrain);
  console.log(`  Rows: ${table.numRows}`);

  const outPath = resolve(GRU_DATA_DIR, "bench-n8n-train.parquet");
  writeParquetFile(table, outPath);
  const stat = Deno.statSync(outPath);
  console.log(`  Written: ${outPath} (${(stat.size / 1e6).toFixed(1)}MB, ${(performance.now() - t0).toFixed(0)}ms)\n`);
}

// ---- Write n8n eval ----
{
  const t0 = performance.now();
  console.log("[Parquet] Building n8n-eval table...");
  const table = buildN8nTable(ds.n8nEval);
  console.log(`  Rows: ${table.numRows}`);

  const outPath = resolve(GRU_DATA_DIR, "bench-n8n-eval.parquet");
  writeParquetFile(table, outPath);
  const stat = Deno.statSync(outPath);
  console.log(`  Written: ${outPath} (${(stat.size / 1e6).toFixed(1)}MB, ${(performance.now() - t0).toFixed(0)}ms)\n`);
}

// ---- Write metadata JSON ----
{
  const metadataPath = resolve(GRU_DATA_DIR, "bench-metadata.json");
  const metadata = {
    leafIds: ds.leafIds,
    embeddingDim: ds.embeddingDim,
    workflowToolLists: ds.workflowToolLists,
    exportedAt: new Date().toISOString(),
    source: "export-to-parquet.ts",
    counts: {
      nodes: ds.nodes.length,
      prodTrain: ds.prodTrain.length,
      prodTest: ds.prodTest.length,
      n8nTrain: ds.n8nTrain.length,
      n8nEval: ds.n8nEval.length,
    },
  };
  Deno.writeTextFileSync(metadataPath, JSON.stringify(metadata));
  const stat = Deno.statSync(metadataPath);
  console.log(`[Metadata] Written: ${metadataPath} (${(stat.size / 1024).toFixed(1)}KB)`);
}

// ---- Summary ----
const mem = (Deno.memoryUsage().rss / 1024 / 1024).toFixed(0);
console.log(`\n=== Export complete (RSS: ${mem}MB) ===`);
console.log("Files:");
console.log("  bench-nodes.parquet");
console.log("  bench-prod-train.parquet");
console.log("  bench-prod-test.parquet");
console.log("  bench-n8n-train.parquet");
console.log("  bench-n8n-eval.parquet");
console.log("  bench-metadata.json");
